homeassistant/components/mqtt/client.py

```diff
@@ -573,16 +673,18 @@ async def _async_perform_subscriptions(
         def _process_client_subscriptions() -> list[tuple[int, int]]:
             """Initiate all subscriptions on the MQTT client and return the results."""
             subscribe_result_list = []
-            for topic, qos in subscriptions:
+            for topic, qos in self._pending_subscriptions.items():
                 result, mid = self._mqttc.subscribe(topic, qos)
                 subscribe_result_list.append((result, mid))
                 _LOGGER.debug("Subscribing to %s, mid: %s, qos: %s", topic, mid, qos)
+            self._pending_subscriptions.clear()
             return subscribe_result_list
 
-        async with self._paho_lock:
-            results = await self.hass.async_add_executor_job(
-                _process_client_subscriptions
-            )
+        async with self._pending_subscriptions_lock:
```

Comment:
Doing this means that, while you are actually doing the subscriptions (which involves doing socket communications, every code path that involves requesting to subscribe blocks until we have finished doing our socket calls.

There is a reason that the original version of this did this:
```

        async with self._subscriptions_lock:
            subscriptions = list(self._pending_subscriptions.items())
            self._pending_subscriptions = {}
        _LOGGER.debug("Subscribing to %s", subscriptions)
```

At the topic of the function.  Because by essentially swapping the array locally (while locked) in one shot, then new subscription requests would not block (an async block is still a block, the code path cannot move forward) waiting for this function to make all it's socket calls.

I recommend you do the swap-to-local trick before ever locking __paho_lock - or you are just going to make subscribes take longer (and the code paths that invoked the subscribes) because they will be backed up waiting on the socket operations (which are supposed to be completely divorced from them - ie. that's the point of a pending queue).

Reply:
done